### PR TITLE
Generate async echo client and server, add async server implementation and basic tests

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

In #1259 code generation was added for async client and servers. We
should generate the echo service and have tests exercising the generated
async code.

Modifications:

- Switch on async code generation for echo
- Implement an async echo service and add some basic tests

Results:

We run tests with an async client and server

### CODE REVIEW

File: Sources/Examples/Echo/Model/echo.grpc.swift [hunk 1]
```diff
@@ -204,7 +299,7 @@ public final class Echo_EchoTestClient: Echo_EchoClientProtocol {
   public func enqueueGetResponse(
     _ response: Echo_EchoResponse,
     _ requestHandler: @escaping (FakeRequestPart<Echo_EchoRequest>) -> () = { _ in }
-  )  {
+  ) {
```
Comment (fabianfett) [inline Sources/Examples/Echo/Model/echo.grpc.swift:302]: nice fix

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      for (i, name) in ["boyle", "jeffers", "holt"].enumerated() {
+        try await update.sendMessage(.with { $0.text = name })
+        let response = try await responseIterator.next()
+        XCTAssertEqual(response, "Swift echo update (\(i)): \(name)")
+      }
+
+      try await update.sendEnd()
+
+      let initialMetadata = try await update.initialMetadata
+      initialMetadata.assertFirst("200", forName: ":status")
```
Comment (fabianfett) [inline Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:136]: Could we check this before the loop already?

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+      let trailingMetadata = try await update.trailingMetadata
+      trailingMetadata.assertFirst("0", forName: "grpc-status")
+
+      let status = await update.status
+      XCTAssertTrue(status.isOk)
```
Comment (fabianfett) [inline Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:142]: We have to `try` the trailingMetadata, but not the status? Is this intended? When do we get the status?

**Review by fabianfett — COMMENTED** [review]
Some questions just to be sure.

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift (hunk 2 above)
Comment (glbrntt) [reply Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:136]: At the moment, yes, we could. However, there's no guarantee that initial metadata is sent back on receipt of request metadata. In particular some gRPC implementations offer RPC implementors a way to set initial metadata which is delayed until the server sends a response message. We don't currently do this but it's been requested and I think we will add it in the future. When that's the case then we can't wait for this before sending the first request.

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift (hunk 3 above)
Comment (glbrntt) [reply Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:142]: It is intended but it's something I'm not sure of API-wise going forward. In the ELF-flavoured API the status promise will always be succeeded. We receive the status at the end of every RPC -- even it doesn't complete (we synthesise one locally if that's the case).

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift (hunk 2 above)
Comment (fabianfett) [reply Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:136]: Maybe we want to add this as a comment?

**Review by fabianfett — APPROVED** [review]

File: Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift (hunk 2 above)
Comment (simonjbeaumont) [reply Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:136]: See #1262 which adds support for our adopters to do just this.

Comment (glbrntt) [reply Tests/GRPCTests/AsyncAwaitSupport/AsyncIntegrationTests.swift:139]: Added a comment in 5be83ddbd4ba5c772b3e6317b84c429b87ce859e

**Pull request merged**